src/text-cairo.c

```diff
@@ -707,11 +707,27 @@ MeasureString (GpGraphics *graphics, GDIPCONST WCHAR *stringUnicode, int *length
 			boundingBox->Width = MaxX;
 			boundingBox->Height = MaxY;
 		}
-		if ((rc->Width > 0) && (boundingBox->Width > rc->Width)) {
-			boundingBox->Width = rc->Width;
+		if (rc->Width > 0) {
+			if (boundingBox->Width > rc->Width) {
+				boundingBox->Width = rc->Width;
+			} else {
+				if (format->formatFlags & StringFormatFlagsDirectionVertical) {
+					boundingBox->Y += (rc->Width - boundingBox->Width) / 2;
+				} else {
+					boundingBox->X += (rc->Width - boundingBox->Width) / 2;
+				}
```

Comment:
I don't see how this makes any sense. Why would you center the bounding box irrespective of the alignment?

Reply:
I wouldn't have noticed any impact on the bounding box directly, because only cairo_MeasureString (and therefore GdipMeasureString) even supply boundingBox to populate, and System.Drawing.Graphics.MeasureString() only uses the size component of that. I was probably trying to fix something but had the wrong place, and the code was never removed – what is here as the final result actually took several iterations to achieve.

The only other thing I can think of is something relating to counting how many characters were fitted. That doesn't make any sense though, because it'll instead mess that up rather badly due to what seems to me to be a bug in that code – it uses boundingBox / rc X + Width, not just Width. Before this change, that would have worked because the only System.Drawing.Graphics.MeasureString overload that returns that information provides a layoutRect (and therefore rc) with X and Y of 0.

So, I would suggest either changing it to take alignment into account and position the bounding box correctly and correcting the character-counting code (technically correct, although probably pointless), or reverting this.

Reply:
I was solving similar issue in the Pango text rendering backend. Ideally we should write some tests for the correct behavior. I ended up returning the actual bounding box of the text with all the alignment taken into account.

The thing is that in the Pango case the line alignment (ie. vertical alignment for most cases) is calculated at the very end once the bounding box height is already known. Unless there is some major flaw in that approach I would prefer to structure the Cairo code in the same way since it makes the code simpler to reason about.

Reply:
Yeah, that's probably the best way. MeasureString is complicated enough (and long enough) that anything that makes it easier to figure out is a good idea.